pkg/broker/registry.go

```diff
@@ -0,0 +1,133 @@
+// Copyright the Service Broker Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package broker
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/gcp-service-broker/brokerapi/brokers/models"
+	"github.com/spf13/viper"
+)
+
+var brokerRegistry = make(map[string]*BrokerService)
+
+func Register(service *BrokerService) {
+	name := service.Name
+
+	if _, ok := brokerRegistry[name]; ok {
+		log.Fatalf("Tried to register multiple instances of: %q", name)
+	}
+
+	brokerRegistry[name] = service
+
+	if err := service.init(); err != nil {
+		log.Fatalf("Error registering service %q, %s", name, err)
+	}
+}
+
+// GetEnabledServices returns a list of all registered brokers that the user
+// has enabled the use of.
+func GetEnabledServices() []*BrokerService {
+	var out []*BrokerService
+
+	for _, svc := range brokerRegistry {
+		if svc.IsEnabled() {
+			out = append(out, svc)
+		}
+	}
+
+	return out
+}
+
+// GetAllServices returns a list of all registered brokers whether or not the
+// user has enabled them.
+func GetAllServices() []*BrokerService {
+	var out []*BrokerService
+
+	for _, svc := range brokerRegistry {
+		out = append(out, svc)
+	}
+
+	return out
+}
+
+type BrokerService struct {
```

Comment:
I'm not a huge fan of this structure where the clients register something that has to be set up with an init() once done, I'd prefer two separate structs but wasn't 100% sure how to get a good solution for two different back-ends (Terraform and native) until I have something to work with and we have the new REST API in.

Reply:
I hear what you're saying. If it helps, the init() style setup you reference seems to be idiomatic with other Go code I've encountered.

Reply:
Oh, I think I understand what you mean a little better. One option is to define a struct abstraction for DefaultServiceDefinition, and then have separate Marshal functions for the different back-ends. It would add some more structure and type safety and make things a little less brittle overall.